### PR TITLE
DOCSUP-14511: Document custom query for database dictionaries

### PR DESCRIPTION
**Pull request by sevirov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:

- Documentation for #26995

### CODE REVIEW

**Comment by sevirov** [comment]
@kitaisreal Please, review this PR.

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 1]
```diff
@@ -464,6 +424,51 @@ LIFETIME(MIN 300 MAX 360)
...
+Setting fields:
+
+-   `db` – Name of the database. Omit it if the database name is set in the `<connection_string>` parameters.
+-   `table` – Name of the table and schema if exists.
+-   `connection_string` – Connection string.
+-   `invalidate_query` – Query for checking the dictionary status. Optional parameter. Read more in the section [Updating dictionaries](../../../sql-reference/dictionaries/external-dictionaries/external-dicts-dict-lifetime.md).
+-   `query` – The custom query. Optional parameter.
+
+!!! info "Note"
+    The `table` and `query` fields cannot be used together. And there must be one of the sources of either `table` or `query`.
```
Comment (gyuton) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:466]:
```suggestion
    The `table` and `query` fields cannot be used together. And either one of the `table` or `query` fields must be declared.
```

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 2]
```diff
@@ -531,6 +538,11 @@ Setting fields:
 
 -   `fail_on_connection_loss` – The configuration parameter that controls behavior of the server on connection loss. If `true`, an exception is thrown immediately if the connection between client and server was lost. If `false`, the ClickHouse server retries to execute the query three times before throwing an exception. Note that retrying leads to increased response times. Default value: `false`.
 
+-   `query` – The custom query. Optional parameter.
+
+!!! info "Note"
+    The `table` or `where` fields cannot be used together with the `query` field. And there must be one of the sources of either `table` or `query`.
```
Comment (gyuton) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:544]:
```suggestion
    The `table` or `where` fields cannot be used together with the `query` field. And either one of the `table` or `query` fields must be declared.
```

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 3]
```diff
@@ -612,6 +628,10 @@ Setting fields:
 -   `where` – The selection criteria. May be omitted.
 -   `invalidate_query` – Query for checking the dictionary status. Optional parameter. Read more in the section [Updating dictionaries](../../../sql-reference/dictionaries/external-dictionaries/external-dicts-dict-lifetime.md).
 -   `secure` - Use ssl for connection.
+-   `query` – The custom query. Optional parameter.
+
+!!! info "Note"
+    The `table` or `where` fields cannot be used together with the `query` field. And there must be one of the sources of either `table` or `query`.
```
Comment (gyuton) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:634]:
```suggestion
    The `table` or `where` fields cannot be used together with the `query` field. And either one of the `table` or `query` fields must be declared.
```

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 4]
```diff
@@ -703,25 +723,28 @@ Example of settings:
...
+-   `column_family` – Name of the column family (table).
+-   `allow_filering` – Flag to allow or not potentially expensive conditions on clustering key columns. Default value is 1.
+-   `partition_key_prefix` – Number of partition key columns in primary key of the Cassandra table. Required for compose key dictionaries. Order of key columns in the dictionary definition must be the same as in Cassandra. Default value is 1 (the first key column is a partition key and other key columns are clustering key).
+-   `consistency` – Consistency level. Possible values: `One`, `Two`, `Three`, `All`, `EachQuorum`, `Quorum`, `LocalQuorum`, `LocalOne`, `Serial`, `LocalSerial`. Default value is `One`.
+-   `where` – Optional selection criteria.
+-   `max_threads` – The maximum number of threads to use for loading data from multiple partitions in compose key dictionaries.
+-   `query` – The custom query. Optional parameter.
+
+!!! info "Note"
+    The `column_family` or `where` fields cannot be used together with the `query` field. And there must be one of the sources of either `column_family` or `query`.
```
Comment (gyuton) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:747]:
```suggestion
    The `column_family` or `where` fields cannot be used together with the `query` field. And either one of the `column_family` or `query` fields must be declared.
```

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 5]
```diff
@@ -764,11 +789,15 @@ Setting fields:
...
+    -   `replica/priority` – The replica priority. When attempting to connect, ClickHouse traverses the replicas in order of priority. The lower the number, the higher the priority.
 -   `db` – Name of the database.
 -   `table` – Name of the table.
--   `where` – The selection criteria. The syntax for conditions is the same as for `WHERE` clause in PostgreSQL, for example, `id > 10 AND id < 20`. Optional parameter.
+-   `where` – The selection criteria. The syntax for conditions is the same as for `WHERE` clause in PostgreSQL. For example, `id > 10 AND id < 20`. Optional parameter.
 -   `invalidate_query` – Query for checking the dictionary status. Optional parameter. Read more in the section [Updating dictionaries](../../../sql-reference/dictionaries/external-dictionaries/external-dicts-dict-lifetime.md).
+-   `query` – The custom query. Optional parameter.
+
+!!! info "Note"
+    The `table` or `where` fields cannot be used together with the `query` field. And there must be one of the sources of either `table` or `query`.
```
Comment (gyuton) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:803]:
```suggestion
    The `table` or `where` fields cannot be used together with the `query` field. And either one of the `table` or `query` fields must be declared.
```

**Review by kitaisreal — APPROVED** [review]

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md [hunk 6]
```diff
@@ -582,6 +596,7 @@ Example of settings:
         <table>ids</table>
         <where>id=10</where>
         <secure>1</secure>
+		<query>SELECT id, value_1, value_2 FROM default.ids</query>
```
Comment (tavplubix) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md:599]: This config seems to be incorrect:
```
Code: 1. DB::Exception: Received from localhost:9000. DB::Exception: Setting `table` or `where` cannot be used with `query` parameter. (UNSUPPORTED_METHOD)

```

**Pull request merged**